### PR TITLE
CU-m9cm8d - Single Transaction Policy CreationUnified policy creation

### PR DESCRIPTION
**Pull request by tgeorgas**

### CODE REVIEW

File: contracts/EntityCoreFacet.sol [hunk 1]
```diff
@@ -51,6 +52,71 @@ import "./Policy.sol";
...
+      _commmissionsBP,
+      _stakeholders 
+    );
+
+    uint256[] memory premiums;
+    uint256 trancheDataLength;
+
+    IPolicy pol = IPolicy(f);
+    uint256 numTranches = _trancheData.length;
+    for (uint i=0; i<numTranches; i++) {
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:97]: Please always use `uint256` instead of `uint`

File: contracts/EntityCoreFacet.sol [hunk 2]
```diff
@@ -51,6 +52,71 @@ import "./Policy.sol";
...
+        premiums[j-2] = _trancheData[i][j];
+      }
+
+      pol.createTranch (
+      _trancheData[i][0], // _numShares
+      _trancheData[i][1], // __pricePerShareAmount
+      premiums
+      );
+    }
+    // uint numTranches = _numShares.length;
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:110]: Please remove all commented out code

File: contracts/EntityCoreFacet.sol [hunk 3]
```diff
@@ -61,15 +127,21 @@ import "./Policy.sol";
...
   {
-    address[] memory stakeholders = new address[](6);
-    stakeholders[0] = address(this);
-    stakeholders[1] = msg.sender;
-    stakeholders[2] = _stakeholders[0];
-    stakeholders[3] = _stakeholders[1];
-    stakeholders[4] = _stakeholders[2];
-    stakeholders[5] = _stakeholders[3];
-
+    address[] memory stakeholders;
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:132]: Code seems redundant - we always pass in stakeholders

File: contracts/base/IEntity.sol [hunk 4]
```diff
@@ -1,4 +1,5 @@
 pragma solidity 0.6.12;
+pragma experimental ABIEncoderV2;
```
Comment (hiddentao) [inline contracts/base/IEntity.sol:2]: Why is this needed?

File: contracts/base/IEntityCoreFacet.sol [hunk 5]
```diff
@@ -12,13 +13,31 @@ interface IEntityCoreFacet {
...
+
+  function createPolicyWithTranches(
+    //Policy Variables
+    uint256[] calldata _dates,
+    address _unit,
+    uint256 _premiumIntervalSeconds,
+    uint256[] calldata _commmissionsBP,
+    address[] calldata _stakeholders,
+
+    //Tranche variable array
```
Comment (hiddentao) [inline contracts/base/IEntityCoreFacet.sol:26]: Please use proper solidity documentation techniques (above function) to document this, not like this

File: contracts/base/IEntityCoreFacet.sol [hunk 6]
```diff
@@ -12,13 +13,31 @@ interface IEntityCoreFacet {
    * @param _commmissionsBP The commissions (basis points, 1 = 0.1%) to pay the broker, capital provider and Nayms
    * @param _stakeholders The three stakeholders of the policy - capital provider, insured party, broker
    */
+
+
+  function createPolicyWithTranches(
+    //Policy Variables
```
Comment (hiddentao) [inline contracts/base/IEntityCoreFacet.sol:19]: Please remove this comment

File: contracts/test/DummyEntityFacet.sol [hunk 7]
```diff
@@ -22,13 +23,29 @@ contract DummyEntityFacet is IDiamondFacet, IEntityCoreFacet {
     return address(this);
   }
 
+  function createPolicyWithTranches(
+    //Policy Variables
```
Comment (hiddentao) [inline contracts/test/DummyEntityFacet.sol:27]: Please remove commented out lines

File: test/policyApprovalsWithUnifiedCall.js [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+//      pricePerShareAmount: [2,3],
+//      premiums: [[10, 20, 30, 40, 50, 60, 70],[1, 2, 3, 4, 5, 6, 7]]
+      trancheData: [[100,2,10, 20, 30, 40, 50, 60, 70],[150,3,1, 2, 3, 4, 5, 6, 7]]
+}, { from: entityManagerAddress })
+
+
+
+
+
+    // const createPolicyTx = await createPolicy(entity, {
```
Comment (hiddentao) [inline test/policyApprovalsWithUnifiedCall.js:152]: Please remove commented out code

File: test/policyApprovalsWithUnifiedCall.js [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+    policyOwnerAddress = entityManagerAddress
+
+    policyProxy = await Policy.at(policyAddress)
+    policy = await IPolicy.at(policyAddress)
+    policyContext = await policyProxy.aclContext()
+
+    // get market address
+    market = await ensureMarketIsDeployed({ artifacts, settings })
+
+    // // setup two tranches
```
Comment (hiddentao) [inline test/policyApprovalsWithUnifiedCall.js:176]: Again

File: test/utils/index.js [hunk 10]
```diff
@@ -157,6 +157,40 @@ export const createEntity = async ({ acl, entityDeployer, adminAddress, entityCo
...
+    brokerCommissionBP = 0,
+    claimsAdminCommissionBP = 0,
+    naymsCommissionBP = 0,
+    underwriter = entity.address,
+    insuredParty = ADDRESS_ZERO,
+    broker = ADDRESS_ZERO,
+    claimsAdmin = ADDRESS_ZERO,
+  } = attrs
+
+  const {
```
Comment (hiddentao) [inline test/utils/index.js:178]: Allow for this to be overridden

File: test/policyApprovalsWithUnifiedCall.js [hunk 11]
```diff
@@ -0,0 +1,405 @@
+import {
```
Comment (hiddentao) [inline test/policyApprovalsWithUnifiedCall.js:1]: Change file name to `policyApprovalsWithUnifiedTranchCreation`

File: contracts/EntityCoreFacet.sol [hunk 12]
```diff
@@ -51,6 +52,71 @@ import "./Policy.sol";
...
+    //Tranche Variable Arrays
+    // uint256[] calldata _numShares,
+    // uint256[] calldata _pricePerShareAmount,
+    // uint256[][] calldata _premiums
+    uint256[][] calldata _trancheData
+  )
+    external
+    override
+  {
+    // address[] memory stakeholders;
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:71]: Pls remove commented out code

File: contracts/EntityCoreFacet.sol [hunk 13]
```diff
@@ -91,6 +163,7 @@ import "./Policy.sol";
     dataBool[__a(pAddr, "isPolicy")] = true; // for _isPolicyCreatedByMe() to work
 
     emit NewPolicy(pAddr, address(this), msg.sender);
+    return pAddr;
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:166]: Add space before this line

File: contracts/EntityCoreFacet.sol [hunk 14]
```diff
@@ -51,6 +52,71 @@ import "./Policy.sol";
...
+    uint256 numTranches = _trancheData.length;
+    for (uint i=0; i<numTranches; i++) {
+      trancheDataLength = _trancheData[i].length;
+      for (uint j=2; j<trancheDataLength; ++j)
+      {
+        premiums[j-2] = _trancheData[i][j];
+      }
+
+      pol.createTranch (
+      _trancheData[i][0], // _numShares
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:105]: Indentation

File: contracts/EntityCoreFacet.sol [hunk 15]
```diff
@@ -51,6 +52,71 @@ import "./Policy.sol";
...
+
+    uint256[] memory premiums;
+    uint256 trancheDataLength;
+
+    IPolicy pol = IPolicy(f);
+    uint256 numTranches = _trancheData.length;
+    for (uint i=0; i<numTranches; i++) {
+      trancheDataLength = _trancheData[i].length;
+      for (uint j=2; j<trancheDataLength; ++j)
+      {
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:100]: Bracket should be on end of prev line, please follow code formatting in rest of file

File: contracts/EntityCoreFacet.sol [hunk 16]
```diff
@@ -61,15 +127,21 @@ import "./Policy.sol";
...
-    stakeholders[2] = _stakeholders[0];
-    stakeholders[3] = _stakeholders[1];
-    stakeholders[4] = _stakeholders[2];
-    stakeholders[5] = _stakeholders[3];
-
+    address[] memory stakeholders;
+    if (_stakeholders[0] == address(this)){
+      stakeholders = _stakeholders;
+    }
+    else{
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:136]: space after else

File: contracts/EntityCoreFacet.sol [hunk 17]
```diff
@@ -61,15 +127,21 @@ import "./Policy.sol";
...
-    address[] memory stakeholders = new address[](6);
-    stakeholders[0] = address(this);
-    stakeholders[1] = msg.sender;
-    stakeholders[2] = _stakeholders[0];
-    stakeholders[3] = _stakeholders[1];
-    stakeholders[4] = _stakeholders[2];
-    stakeholders[5] = _stakeholders[3];
-
+    address[] memory stakeholders;
+    if (_stakeholders[0] == address(this)){
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:133]: Space before bracket, please follow code formatting elsewhere

File: contracts/test/DummyEntityFacet.sol (hunk 7 above)
Comment (hiddentao) [reply contracts/test/DummyEntityFacet.sol:27]: Space after //

**Review by hiddentao — CHANGES_REQUESTED** [review]
Please:

* Get CI passing.
* Use the clickup id in atleast one of the commits so that linkage takes place. 
* Update PR title to include clickup id (look at other PRs)

File: contracts/base/IEntity.sol (hunk 4 above)
Comment (najienka) [reply contracts/base/IEntity.sol:2]: ABIEncoderV2 enables passing of struct data types into a function from web3 or another contract.
If this is not needed, we can remove.

e.g., 
struct Data {
    string text;
  }

function Insert(Data memory recordData) public returns(bool success) {}

**Review by najienka — CHANGES_REQUESTED** [review]
Added a comment on `ABIEncoderV2` under file: contracts/base/IEntity.sol and agree with the other comments.

**Comment by hiddentao** [comment]
@tgeorgas Pls resolve these issues

Also, please add CU ticket id to PR name (see other PRs for example)

**Comment by melks** [comment]
Task linked: [CU-m9cm8d Single Transaction Policy Creation on Smart Contracts (solidity)](https://app.clickup.com/t/m9cm8d)

File: contracts/base/IEntityCoreFacet.sol [hunk 18]
```diff
@@ -12,13 +13,21 @@ interface IEntityCoreFacet {
    * @param _commmissionsBP The commissions (basis points, 1 = 0.1%) to pay the broker, capital provider and Nayms
    * @param _stakeholders The three stakeholders of the policy - capital provider, insured party, broker
    */
+
```
Comment (hiddentao) [inline contracts/base/IEntityCoreFacet.sol:16]: Pls fix the function docs here. Pls remove all the comments inside the argument list both there and everywhere else. Arguments should be documented in the function block documentation as shown in the original version of the file.

File: contracts/EntityCoreFacet.sol [hunk 19]
```diff
@@ -14,175 +15,192 @@ import "./Policy.sol";
...
+
+    /**
+     * Constructor
+     */
+    constructor(address _settings) public Controller(_settings) {}
+
+    // IDiamondFacet
+    function getSelectors() public pure override returns (bytes memory) {
+        return
+            abi.encodePacked(
```
Comment (hiddentao) [inline contracts/EntityCoreFacet.sol:48]: These changes are just about formatting and indentation. Pls fix to use 2 spaces for indents everywhere.

File: test/utils/index.js [hunk 20]
```diff
@@ -157,48 +157,71 @@ export const createEntity = async ({ acl, entityDeployer, adminAddress, entityCo
   return entityAddress
 }
 
+//export const createPolicy = async (entity, attrs = {}, ...callAttrs) => {
```
Comment (hiddentao) [inline test/utils/index.js:160]: Pls remove comment

File: test/utils/index.js [hunk 21]
```diff
@@ -157,48 +157,71 @@ export const createEntity = async ({ acl, entityDeployer, adminAddress, entityCo
...
 export const createPolicy = async (entity, attrs = {}, ...callAttrs) => {
   const currentTime = ~~(Date.now() / 1000)
+  var ret = '';
```
Comment (hiddentao) [inline test/utils/index.js:163]: use `let` instead of `var`

File: test/utils/index.js [hunk 22]
```diff
@@ -157,48 +157,71 @@ export const createEntity = async ({ acl, entityDeployer, adminAddress, entityCo
...
+    // underwriter = entity.address,
     claimsAdmin = ADDRESS_ZERO,
+    insuredParty = ADDRESS_ZERO
   } = attrs
 
+  var trancheData = attrs.trancheData
+
+  var underwriter = attrs.underwriter
+
+  if(underwriter == undefined){
```
Comment (hiddentao) [inline test/utils/index.js:188]: just use `(!underwriter)`

File: test/utils/index.js [hunk 23]
```diff
@@ -157,48 +157,71 @@ export const createEntity = async ({ acl, entityDeployer, adminAddress, entityCo
...
-    underwriter = entity.address,
-    insuredParty = ADDRESS_ZERO,
+
     broker = ADDRESS_ZERO,
+    // underwriter = entity.address,
     claimsAdmin = ADDRESS_ZERO,
+    insuredParty = ADDRESS_ZERO
   } = attrs
 
+  var trancheData = attrs.trancheData
```
Comment (hiddentao) [inline test/utils/index.js:184]: use `let` instead of `var`

File: test/utils/index.js [hunk 24]
```diff
@@ -157,48 +157,71 @@ export const createEntity = async ({ acl, entityDeployer, adminAddress, entityCo
...
+
+  var underwriter = attrs.underwriter
+
+  if(underwriter == undefined){
+    underwriter = entity.address;
+  }
+
+  //Must pass empty trancheData if not creating tranches
+  if (trancheData === undefined) {
```
Comment (hiddentao) [inline test/utils/index.js:193]: use `!tranchData`

Comment (hiddentao) [reply test/utils/index.js:193]: `tranchDate = tranchData || []`

**Comment by Huanzhang89** [comment]
Reopened in https://github.com/nayms/contracts/pull/110 with correctly named branch
